src/main/java/com/adroll/cantor/HLLCounter.java

```diff
@@ -490,41 +453,48 @@ public static long intersect(List<HLLCounter> hs) {
       return 0;
     }
     for (HLLCounter hll : hs) {
-      if (hll.size() == 0) {
+      if ((hll.getMinHash() == null && hll.size() == 0) || hll.getMinHash().size() == 0) {
         return 0;
       }
     }
-    TreeSet<Long> all = new TreeSet<Long>();
     int mink = Integer.MAX_VALUE;
     int maxs = Integer.MIN_VALUE;
+    ArrayList<Long[]> minhash_arrays = new ArrayList();
     for(HLLCounter h : hs) {
       if(h.isIntersectable()) {
-        all.addAll(h.getMinHash());
+        minhash_arrays.add(h.getMinHash().toArray(new Long[0]));
         mink = Math.min(mink, h.getK());
         maxs = Math.max(maxs, h.getMinHash().size());
       }
     }
+    // The indices in the different minhashes where we are checking
+    // if the values match
+    int[] c_idx = new int[minhash_arrays.size()];
     mink = maxs < mink ? maxs : mink;
     int result = 0;
-    for(int i = 0; i < mink; i++) {
-      long l = 0;
-      try {
-        l = all.pollFirst();
-      } catch(NullPointerException e) {
-        //This can happen if k is larger than
-        //the number of insertions.
-        break;
-      }
-      boolean allContain = true;
-      for(HLLCounter h : hs) {
-        if(h.isIntersectable()) {
-          if(!h.getMinHash().contains(l)) {
-            allContain = false;
-            break;
-          }
+    long min;
+    boolean inAllSets = true;
+    boolean changed = false;
+    for (int k = 0; k < mink; k++) {
```

Comment:
been staring at this for loop for way too long, I'm still not sure exactly what a lot of these variables are tracking. I think it would be good for us to just sit down and go through it.